### PR TITLE
Consul EDS Blocking Queries

### PR DESCRIPTION
**Pull request by kdorosh**
# Description

Consul EDS discovery query mode is now configurable (blocking vs non-blocking consul queries) via `eds_blocking_queries` on the Gloo `Settings`. The default remains the same (non-blocking queries) becuase this is more performant with cache hits on the consul agent (caching also enabled by default) but can be more performant when caching is disabled since blocking queries should result in fewer network calls.

# Context

Adding features to improve performance with consul at scale.

# Checklist:

- [ ] I included a concise, user-facing changelog (for details, see https://github.com/solo-io/go-utils/tree/master/changelogutils) which references the issue that is resolved.
- [ ] If I updated APIs (our protos) or helm values, I ran `make -B install-go-tools generated-code` to ensure there will be no code diff
- [ ] I followed guidelines laid out in the Gloo Edge [contribution guide](https://docs.solo.io/gloo-edge/latest/contributing/)
- [ ] I opened a draft PR or added the work in progress label if my PR is not ready for review
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

BOT NOTES: 
resolves https://github.com/solo-io/gloo/issues/6599
resolves https://github.com/solo-io/gloo/issues/7037

### CODE REVIEW

File: test/e2e/consul_test.go [hunk 1]
```diff
@@ -126,145 +87,266 @@ var _ = Describe("Consul e2e", func() {
...
 			Expect(err).NotTo(HaveOccurred())
 		}
-		envoyInstance.Clean()
 		cancel()
 	})
 
-	It("works as expected", func() {
-		_, err := testClients.ProxyClient.Write(getProxyWithConsulRoute(writeNamespace, envoyPort), clients.WriteOpts{Ctx: ctx})
-		Expect(err).NotTo(HaveOccurred())
+	Context("with envoy and gloo", func() {
```
Comment (kdorosh) [inline test/e2e/consul_test.go:93]: note for reviewer; this PR adds one more e2e test at the end of the file, the rest of the diff is caused by moving the existing tests into this new `Context` for "with envoy and gloo"

**Comment by kdorosh** [comment]
/kick should work with ssl flake https://storage.googleapis.com/solo-public-build-logs/logs.html?buildid=ac256f82-c2fe-4cce-b789-4ea7c42f6aa3

File: test/e2e/consul_test.go (hunk 1 above)
Comment (kdorosh) [reply test/e2e/consul_test.go:93]: also the EDS e2e test was rewritten to run both with blocking and non-blocking eds queries

**Comment by kdorosh** [comment]
/kick flake? https://storage.googleapis.com/solo-public-build-logs/logs.html?buildid=3f6f857f-170a-48f3-b317-bcc2eceff8f3

File: projects/gloo/pkg/plugins/aws/ec2/eds.go [hunk 2]
```diff
@@ -18,7 +18,7 @@ import (
 
 // EDS API
 // start the EDS watch which sends a new list of endpoints on any change
-func (p *plugin) WatchEndpoints(writeNamespace string, unfilteredUpstreams v1.UpstreamList, opts clients.WatchOpts) (<-chan v1.EndpointList, <-chan error, error) {
+func (p *plugin) WatchEndpoints(writeNamespace string, unfilteredUpstreams v1.UpstreamList, opts clients.WatchOpts, settings *v1.Settings) (<-chan v1.EndpointList, <-chan error, error) {
```
Comment (ianmacclancy) [inline projects/gloo/pkg/plugins/aws/ec2/eds.go:21]: Settings is taken as a parameter but not used, there is also no need to take in settings as a parameter as the plugin has the settings if init is correctly run. you can access them through p.settings

File: projects/gloo/pkg/plugins/kubernetes/eds.go [hunk 3]
```diff
@@ -23,7 +23,7 @@ import (
 	"k8s.io/apimachinery/pkg/labels"
 )
 
-func (p *plugin) WatchEndpoints(writeNamespace string, upstreamsToTrack v1.UpstreamList, opts clients.WatchOpts) (<-chan v1.EndpointList, <-chan error, error) {
+func (p *plugin) WatchEndpoints(writeNamespace string, upstreamsToTrack v1.UpstreamList, opts clients.WatchOpts, settings *v1.Settings) (<-chan v1.EndpointList, <-chan error, error) {
```
Comment (ianmacclancy) [inline projects/gloo/pkg/plugins/kubernetes/eds.go:26]: Same comment as above - 

Settings is taken as a parameter but not used, there is also no need to take in settings as a parameter as the plugin has the settings if init is correctly run. you can access them through p.settings

**Review by ianmacclancy — CHANGES_REQUESTED** [review]

File: projects/gloo/pkg/plugins/kubernetes/eds.go (hunk 3 above)
Comment (kdorosh) [reply projects/gloo/pkg/plugins/kubernetes/eds.go]: good call, updated

File: projects/gloo/pkg/plugins/aws/ec2/eds.go (hunk 2 above)
Comment (kdorosh) [reply projects/gloo/pkg/plugins/aws/ec2/eds.go]: good call, updated

**Comment by kdorosh** [comment]
/kick flake? passes locally reliably https://storage.googleapis.com/solo-public-build-logs/logs.html?buildid=b5ef9f1f-37eb-4a2a-8c14-7b5a52988d8b

File: projects/gloo/api/v1/settings.proto [hunk 4]
```diff
@@ -319,6 +319,17 @@ message Settings {
...
+        // Enables blocking queries for Gloo's requests to the Consul Catalog API for each service
+        // (`/catalog/service/:servicename`) to get endpoints for EDS. For more on blocking queries, see
+        // https://www.consul.io/api-docs/features/blocking
+        //
+        // Enabling this feature will likely result in fewer network calls to Consul, but may also result in
+        // fewer local consul agent cache hits for Gloo's requests to the Consul Catalog API. (see
+        // `query_options` above to configure caching; caching is enabled by default).
+        //
+        // Defaults to false.
+        google.protobuf.BoolValue eds_blocking_queries = 23;
```
Comment (ianmacclancy) [inline projects/gloo/api/v1/settings.proto:332]: Where is this setting value used?

Comment (kdorosh) [reply projects/gloo/api/v1/settings.proto:332]: gets used here https://github.com/solo-io/gloo/blob/cba7ba77165c8927a56517093cd851f25fee79d6/projects/gloo/pkg/plugins/consul/eds.go#L121

**Review by ianmacclancy — DISMISSED** [review]

**Review by elcasteel — DISMISSED** [review]

File: projects/gloo/pkg/plugins/consul/eds.go [hunk 5]
```diff
@@ -157,14 +158,14 @@ func (p *plugin) WatchEndpoints(writeNamespace string, upstreamsToTrack v1.Upstr
 				// in testing, I only saw cache hits here if last index was zero (first call during blocking)
 
 				// construct a set of the present services by datacenter
-				dcToCurrentSvcs := map[string]map[string]struct{}{}
+				dcToCurrentSvcs := map[string]sets.String{}
```
Comment (elcasteel) [inline projects/gloo/pkg/plugins/consul/eds.go:161]: This makes more sense.

**Review by elcasteel — APPROVED** [review]

**Pull request merged**